tests/rptest/clients/kcl.py

```diff
@@ -273,35 +274,56 @@ def alter_partition_reassignments(self,
         no_broker_re = re.compile(
             r"^(?P<topic>[a-z\-]+?) +(?P<partition>[0-9]+?) +BROKER_NOT_AVAILABLE.*$"
         )
-        bad_rep_factor_re = re.compile(
-            r"^(?P<topic>[a-z\-]+?) +(?P<partition>[0-9]+?) +INVALID_REPLICATION_FACTOR.*$"
+        no_topic_or_part_re = re.compile(
+            r"^(?P<topic>[a-z\-]+?) +(?P<partition>[0-9]+?) +UNKNOWN_TOPIC_OR_PARTITION.*$"
         )
-        for l in lines:
-            l = l.strip()
-            self._redpanda.logger.debug(l)
 
-            m = no_broker_re.match(l)
-            # No broker available means the partition did not find any eligible allocation nodes.
-            # See the map from cluster errors to kafka errors in kafka::map_topic_error_code()
-            if m is not None:
-                raise RuntimeError('No eligible allocation nodes')
+        # Unkown topic or partition occurs: (1) when the request was sent to a broker with
+        # out-of-date metadata (i.e., the broker does not yet know the topic exists) or
+        # (2) when the number of replicas for one (or more) partitions in a request does
+        # not match the topic replication factor.
 
-            m = bad_rep_factor_re.match(l)
-            # Invalid replication factor means the number of replicas for one (or more) partitions
-            # in a request does not match the replication factor for the topic.
-            if m is not None:
-                raise RuntimeError(
-                    'Number of replicas != topic replication factor')
+        # We have a small retry loop here for point (1) because the broker should get
+        # updated metadata soon. If the loop still fails, then point (2) is likely
+        # the reason.
+        lines = None
+
+        def do_alter_partitions():
+            nonlocal lines
+            lines = self._cmd(cmd).splitlines()
+
+            for l in lines:
+                l = l.strip()
+                self._redpanda.logger.debug(l)
+
+                m = no_broker_re.match(l)
+                # No broker available means the partition did not find any eligible allocation nodes.
+                # See the map from cluster errors to kafka errors in kafka::map_topic_error_code()
+                # Do not retry for this error/
+                if m is not None:
+                    raise RuntimeError('No eligible allocation nodes')
+
+                m = no_topic_or_part_re.match(l)
+                # Retry on unknown topic or partition
+                if m is not None:
+                    self._redpanda.logger.debug(
+                        'Number of replicas != topic replication factor')
+                    return False
+
+                # The line should be OK if no other errors occurred
+                assert ok_re.match(l) is not None
+
+            return True
 
-            m = ok_re.match(l)
-            assert m is not None
+        wait_until(do_alter_partitions,
+                   timeout_sec=5,
```

Comment:
this sort of retry should probably be in your test, not the client. for example, if you ran alter partitions on a topic that you knew did not exist, would you still pay the cost of the wait_until loop? that is, it seems like behavior specific to your test, rather than behavior desired by all users of the client.

Reply:
It is specific to the test. I moved this code to partition_reassignments_test.py

Reply:
On second thought,
https://github.com/redpanda-data/redpanda/pull/8849#discussion_r1114733798